# TechnoHacks Edutch Internship Task 1

## Objective: To remove missing data and outliers

## Given Dataset: The Iconic Titanic Dataset from Kaggle

In [ ]:
Name of File to be worked upon: train.csv
Basic details and understanding of Dataset.
Number of columns: 12 
Column Description:
    Passenger ID : Unique id of the passenger
    Survived: To note weather the passenger survived or not. 0 for Not Servived and 1 indicates Survived.
    Pclass: to indicate the class of passenger ticket. 1 means upper class, 2 means middle class and 3 means lower class.
    Name : Name of the passenger who purchased  the ticket and his family members if any.
    Sex : Sex of passenger. Male or Female.
    Age : Age of passenger.
    Sibsp: Indicates weather passenger is travelling with his siblings or spouse. if Yes then 1 orelse 0.
    Parch: Denotes weather parents are present.
    Ticket: The ticket number of the passenger
    Fare: The total fare of the passenger for that particular class of ticket 
    Cabin: The cabin number alloted to that particular passenger
    Embarked: The place from where the passenger embarked on the journey.C = Cherbourg, Q = Queenstown, S = Southampton

In [3]:
#Process to achieve the objective

## Importing the data

In [4]:
#installing pandas in the system.
# importing pandas library 
# Using the pandas library we can import the csv file named train.csv.
# Then create a dataframe so that operations can be performed on the given data.
#storing the data from the csv file in the dataframe.

In [5]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.


In [5]:
import pandas as pd
import numpy as np

In [6]:
df = pd.read_csv('train.csv')

In [7]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

## Basic Data analysis

In [8]:
# using the pandas we can see the shape of the dataset
# we can see the 1st 5 records using the head()
# we can look at the last 5 records using the tail()

In [9]:
df.shape

(891, 12)

Indicates that the given dataset contains 891 rows and 12 columns

In [11]:
df.head() # returns the 1st 5 records of the dataset

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [12]:
df.tail() # returns the last 5 records of the dataset.

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [13]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

## Exploratory Data analysis

In [14]:
# 1] to perform Basic statiscal analysis on the given datasset.

In [15]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

## Insights

1. Total count of passengers is 891
2. Average age of the passengers is 29.6991
3. Rate of Passengers with siblings/spouse is 0.523
4. Minimum Fare paid = 0
5. Maximum Fare paid = 512.32
6. Average fare = 32.20
7. Max age of passenger = 80 years
8. Min age of passenger = 0.42 years

In [16]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

Shows the name of all the columns present in the Dataset.

In [17]:
df[['Name','Pclass','Sex','Age']] #shows the records of particular columns

Name  Pclass     Sex   Age
0                              Braund, Mr. Owen Harris       3    male  22.0
1    Cumings, Mrs. John Bradley (Florence Briggs Th...       1  female  38.0
2                               Heikkinen, Miss. Laina       3  female  26.0
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)       1  female  35.0
4                             Allen, Mr. William Henry       3    male  35.0
..                                                 ...     ...     ...   ...
886                              Montvila, Rev. Juozas       2    male  27.0
887                       Graham, Miss. Margaret Edith       1  female  19.0
888           Johnston, Miss. Catherine Helen "Carrie"       3  female   NaN
889                              Behr, Mr. Karl Howell       1    male  26.0
890                                Dooley, Mr. Patrick       3    male  32.0

[891 rows x 4 columns]

## Missing Values

What is Missing Values?
    Missing values are those values in a particular record that have not been recorded or are not available due to human or
    mechanical error.

How to handle Missing Values?
    There are multiple ways of Handling missing Values such as 
    1. Removing the particular record rows
    2. Subsituting the value with the average value
    3. Filling the value with the most frequent value.

### Process to handle Missing Data

We will first identify the count of missing data in each column 
Sum up the count of missing data
Drop the rows containing missing data

In [18]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [19]:
df2 = df.dropna()

In [20]:
df2

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
10            11         1       3   
11            12         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                           Carlsson, Mr. Frans Olof    male  33.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
1        0  PC 17599  71.2833          C85        C  
3        0    113803  53.1000         C123        S  
6        0     17463  51.8625          E46        S  
10       1   PP 9549  16.7000           G6        S  
11       0    113783  26.5500         C103        S  
..     ...       ...      ...          ...      ...  
871      1     11751  52.5542          D35        S  
872      0       695   5.0000  B51 B53 B55        S  
879      1     11767  83.1583          C50        C  
887      0    112053  30.0000          B42        S  
889      0    111369  30.0000         C148        C  

[183 rows x 12 columns]

In [21]:
df2.head(10)

PassengerId  Survived  Pclass  \
1             2         1       1   
3             4         1       1   
6             7         0       1   
10           11         1       3   
11           12         1       1   
21           22         1       2   
23           24         1       1   
27           28         0       1   
52           53         1       1   
54           55         0       1   

                                                 Name     Sex   Age  SibSp  \
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
21                              Beesley, Mr. Lawrence    male  34.0      0   
23                       Sloper, Mr. William Thompson    male  28.0      0   
27                     Fortune, Mr. Charles Alexander    male  19.0      3   
52           Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.0      1   
54                     Ostby, Mr. Engelhart Cornelius    male  65.0      0   

    Parch    Ticket      Fare        Cabin Embarked  
1       0  PC 17599   71.2833          C85        C  
3       0    113803   53.1000         C123        S  
6       0     17463   51.8625          E46        S  
10      1   PP 9549   16.7000           G6        S  
11      0    113783   26.5500         C103        S  
21      0    248698   13.0000          D56        S  
23      0    113788   35.5000           A6        S  
27      2     19950  263.0000  C23 C25 C27        S  
52      0  PC 17572   76.7292          D33        C  
54      1    113509   61.9792          B30        C

In [22]:
df2.tail()

PassengerId  Survived  Pclass  \
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                 Name     Sex   Age  SibSp  \
871  Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                          Carlsson, Mr. Frans Olof    male  33.0      0   
879     Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                      Graham, Miss. Margaret Edith  female  19.0      0   
889                             Behr, Mr. Karl Howell    male  26.0      0   

     Parch  Ticket     Fare        Cabin Embarked  
871      1   11751  52.5542          D35        S  
872      0     695   5.0000  B51 B53 B55        S  
879      1   11767  83.1583          C50        C  
887      0  112053  30.0000          B42        S  
889      0  111369  30.0000         C148        C

## Removing Duplicates 

In [23]:
df2 = df2['Name'].drop_duplicates()

In [26]:
df2.describe()

count                                                   183
unique                                                  183
top       Cumings, Mrs. John Bradley (Florence Briggs Th...
freq                                                      1
Name: Name, dtype: object

# Outliers

## Outliers are those values in the dataset that are at very high deviation from the mean value and can cause significant disruption in the analysis of data 

In [66]:
# The outliers in our dataset can be present in columns whose datatypes are int64 or float64

In [69]:
df2.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [72]:
#to check outliers in the age column we use the Z score method and the formula Z = (i-mean)/standard deviation.
age = []
for i in df2['Age']:
    age.append(i)

# We use Z Score method to find the outliers

In [77]:
outliers =[]
def detect_outliers(data):
    
    threshold = 3
    mean = np.mean(data)
    std = np.std(data)
    
    for i in data:
        z_score = (i - mean)/ std
        if np.abs(z_score) > threshold:
            outliers.append(i)
    return outliers

In [78]:
outlier_pt = detect_outliers(age)

In [79]:
outlier_pt

[]

## Using the above method we did not find any outlier in the Age column.

In [80]:
#to check outliers in the Fare column we use the Z score method and the formula Z = (i-mean)/standard deviation.
ticket_price = []
for i in df2['Fare']:
    ticket_price.append(i)

In [81]:
outlier_fare = detect_outliers(ticket_price)

In [82]:
outlier_fare

[512.3292, 512.3292]

In [83]:
#As the fare column contains 2 outliers we will remove the outliers to clean the data

In [93]:
df2.tail()

PassengerId  Survived  Pclass  \
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                 Name     Sex   Age  SibSp  \
871  Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                          Carlsson, Mr. Frans Olof    male  33.0      0   
879     Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                      Graham, Miss. Margaret Edith  female  19.0      0   
889                             Behr, Mr. Karl Howell    male  26.0      0   

     Parch  Ticket     Fare        Cabin Embarked  
871      1   11751  52.5542          D35        S  
872      0     695   5.0000  B51 B53 B55        S  
879      1   11767  83.1583          C50        C  
887      0  112053  30.0000          B42        S  
889      0  111369  30.0000         C148        C